django/contrib/auth/hashers.py

```diff
@@ -703,3 +703,60 @@ def safe_summary(self, encoded):
 
     def harden_runtime(self, password, encoded):
         pass
+
+
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
+
+    def encode(self, password, salt, n=16384, r=8, p=1):
+        assert password is not None
+        assert salt and '$' not in salt
+        hash = scrypt(password, salt=salt, n=n, r=r, p=p)
+        hash = base64.b64encode(hash).decode('ascii').strip()
+        return "%s$%d$%s$%s$%d$%d" % (self.algorithm, n, salt, hash, r, p)
+
+    def decode(self, encoded):
+        algorithm, work_factor, salt, hash, block_size, parallelism = encoded.split('$', 5)
+        assert algorithm == self.algorithm
+        return {
+            'algorithm': algorithm,
+            'hash': hash,
+            'work_factor': int(work_factor),
+            'salt': salt,
+            'block_size': int(block_size),
+            'parallelism': int(parallelism)
+        }
+
+    def verify(self, password, encoded):
+        decoded = self.decode(encoded)
+        encoded_2 = self.encode(password, decoded['salt'], decoded['work_factor'],
+                                decoded['block_size'], decoded['parallelism'])
+        return constant_time_compare(encoded, encoded_2)
+
+    def safe_summary(self, encoded):
+        decoded = self.decode(encoded)
+        return {
+            _('algorithm'): decoded['algorithm'],
+            _('work_factor'): decoded['work_factor'],
+            _('salt'): mask_hash(decoded['salt']),
+            _('hash'): mask_hash(decoded['hash']),
+        }
+
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return decoded['work_factor'] != self.work_factor
+
+    def harden_runtime(self, password, encoded):
+        decoded = self.decode(encoded)
+        extra_iterations = self.work_factor - decoded['work_factor']
+        if extra_iterations > 0:
```

Comment:
What about `r` and `p` here?

Reply:
I thought the harden_runtime function was based only off of the work_factor, according to lines 235-236 in the hashers.py file.

Reply:
I "depends" :) `harden_runtime` is there to bridge the CPU timing difference between the old and new settings, "work factor" in that comment might have been the wrong wording due to the fact that hasher like `scrypt` actually have a parameter called similar. So the main question here is whether `r` and `p` contribute to the runtime in a way that the should be included here (and how, ie linear or quadratic). According to https://blog.filippo.io/the-scrypt-parameters/ it says the following for `r`: 
> It does that by iterating the hash function 2𝑟 times, so both memory usage (to store the hash values) and CPU time scale linearly with it. That is, if 𝑟 doubles the resources double.

For `p` it says:
> 𝑝 > 1 can be handled in two ways: sequentially, which does not increase memory usage but requires 𝑝 times the CPU and wall clock time; or parallelly, which requires 𝑝 times the memory and effective CPU time, but does not increase wall clock time.

So it will be hard to include `p` in `harden_runtime` (I guess we can ignore it and pretend it does not change the wall time because the server has enough CPU -- probably not true but including the number of available CPU threads in this calculation seems overkill).

Reply:
Okay I see. I added the block_size to the runtime but left the parallelism out. Thanks for your help.